### PR TITLE
76 flooding direct damages

### PR DESCRIPTION
**Pull request by thomas-fred**
- Damage curves, categorised by hazard type (flood, storm, earthquake, etc.), sub-categorised by asset type (road bridge, unpaved road segment, railway etc.)
- Calculation of damage fraction in range [0, 1] for exposed edge segments
- Plotting of damage fraction distributions
- Calculation of direct damages for each segment of edge split by raster, then collapsed back to original (unsplit) edges
- Calculation of Expected Annual Damages (integrate out probability from per return period costs).
- Aggregation of results to administrative boundaries.

To generate a geoparquet file of estimated direct damage costs:
```
snakemake --cores all results/egypt-latest_filter-road/hazard-aqueduct-river/damage_EAD_and_cost_per_RP.geoparquet
```

To plot the distributions of damage fractions:
```
snakemake --cores 1 results/egypt-latest_filter-road/hazard-aqueduct-river/damage_fraction_plots
```
Which produces this [set of histograms](https://github.com/nismod/open-gira/files/9802641/road_unpaved_damage_fraction.pdf).

To estimate expected annual damages in dollars:
```
snakemake --cores all results/egypt-latest_filter-road/hazard-aqueduct-river/damage_EAD_and_cost_per_RP.geoparquet
```

To aggregate expected annual damages to a set of administrative regions:
```
snakemake --cores all results/egypt-latest_filter-road/hazard-aqueduct-river/EAD_and_cost_per_RP/agg-sum/admin-level-0.geoparquet
```

### CODE REVIEW

File: workflow/scripts/transport/direct_damages.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            .multiply(damage_fraction[SPLIT_LENGTH_FIELD], axis="index")
+
+    logging.info(f"Reading raw network data for unsplit geometry")
+    unsplit: gpd.GeoDataFrame = gpd.read_parquet(unsplit_path)
+    logging.info(f"{unsplit.shape=}")
+
+    # join the other fields with the direct damage estimates
+    logging.info("Unifying rasterised segments and summing damage costs")
+    direct_damages: gpd.GeoDataFrame = gpd.GeoDataFrame(
+        pd.concat(
```
Comment (tomalrussell) [inline workflow/scripts/transport/direct_damages.py:193]: This concat feels a bit fragile, though I can convince myself it's right. What do you think about something like the below?

- stick an `edge_id` column back on `direct_damages_only` and `.groupby('edge_id').sum()`
- pick all `non_hazard_columns` (including `geometry`) from `unsplit` and set `edge_id` as index
- join the two

```python
grouped_direct_damages = pd.concat(
    [direct_damages_only, damage_fraction["edge_id"], 
    axis="columns"
).set_index("edge_id")
unsplit_subset = unsplit[non_hazard_columns].set_index("edge_id")
direct_damages = unsplit_subset.join(grouped_direct_damages, validate='one_to_one')
```

**Review by tomalrussell — APPROVED** [review]
Tidy handling of hazard intensity / damage curves and damage calculations, broadly looks solid 💯

File: workflow/rules/preprocess/create_overall_bbox.smk [hunk 2]
```diff
@@ -3,10 +3,8 @@ rule create_overall_bbox:
     conda: "../../../environment.yml"
     input:
         "{OUTPUT_DIR}/input/{DATASET}.osm.pbf",
```
Comment (tomalrussell) [inline workflow/rules/preprocess/create_overall_bbox.smk:5]: `{DATASET}_filter-{FILTER}.osm.pbf`?

Comment (thomas-fred) [reply workflow/rules/preprocess/create_overall_bbox.smk:5]: I didn't think I'd changed the naming scheme with regard to filtering. What would be rationale be for this change?

File: workflow/scripts/transport/direct_damages.py (hunk 1 above)
Comment (thomas-fred) [reply workflow/scripts/transport/direct_damages.py:193]: There were _reasons_ why I did it this way, but yes I agree relying on the repeated index values seems more brittle than using the `edge_id`.

I think you're missing a call to groupby in the above block? I'll see if I can get something like your suggestion to work.

File: workflow/scripts/transport/create_rail_network.py [hunk 3]
```diff
@@ -85,20 +101,26 @@ def get_rehab_costs(row: pd.Series, rehab_costs: pd.DataFrame) -> Tuple[float, f
     try:
         edges = gpd.read_parquet(osm_edges_path)
     except ValueError as error:
-        # if the input parquet file does not contain a geometry column, geopandas
-        # will raise a ValueError rather than try to procede
-        logging.info(f"{error}\n" "writing empty files and skipping processing...")
-        write_empty_frames(edges_output_path, nodes_output_path)
-        sys.exit(0)  # exit gracefully so snakemake will continue
+        if NO_GEOM_ERROR_MSG in str(error):
```
[the remark this answers is not in the document]
Comment (thomas-fred) [reply workflow/scripts/transport/create_rail_network.py:104]: Yeah I think that's a good idea. However I'm not sure how to do that given the current directory structure -- I don't think I've successfully used relative imports from parent modules, for instance. Perhaps including a setup.py and installing workflow/scripts as a package is the way to go -- then we can use imports relative to the package root, right?

[the remark this answers is not in the document]
Comment (tomalrussell) [reply workflow/scripts/transport/create_rail_network.py:104]: Yeah, I think a package may well be the way to go - I'd lean towards pulling it out of the `workflow/scripts` tree entirely and giving it a name, then we can `pip install {-e} ./src/open_gira` and `import helper_function from open_gira.utils`

File: workflow/rules/preprocess/create_overall_bbox.smk (hunk 2 above)
Comment (tomalrussell) [reply workflow/rules/preprocess/create_overall_bbox.smk:5]: [pasting from orphaned comment] Ah, this was my quick guess when I saw the tests failing but you were still pushing commits. Feel free to ignore.

**Pull request merged**